File: lib/player/stream_url_classifier.dart
/// URL classification aligned with `/player/StreamUrlClassifier.kt` (EaMax gateway pages).
class StreamUrlClassifier {
  StreamUrlClassifier._();

  /// `.php` gateways — include `.php/` paths (many panels use `/play.php/channel`).
  static bool isPhpLikeUrl(String? url) {
    if (url == null || url.trim().isEmpty) return false;
    final u = url.toLowerCase();
    return RegExp(r'\.php(\$|[/?#])', caseSensitive: false).hasMatch(u);
  }

  /// PHP, ASP, /player/, /embed/ gateways — need in-app web player, not direct Exo.
  static bool isLikelyGatewayUrl(String? url) {
    if (url == null || url.trim().isEmpty) return false;
    final u = url.toLowerCase();
    return RegExp(r'\.(php|asp|aspx|cgi|jsp)(\?|#|\$|/)', caseSensitive: false).hasMatch(u) ||
        u.contains('/embed/') ||
        u.contains('/gateway/') ||
        u.contains('/stream/') ||
        u.contains('/play/') ||
        u.contains('/player/');
  }

  static bool needsWebPlayer(String url) {
    final u = url.trim();
    if (u.isEmpty) return false;
    final hasDirect = hasObviousM3u8(u) || hasObviousMpd(u) || hasObviousTs(u);
    return (isPhpLikeUrl(u) || isLikelyGatewayUrl(u)) && !hasDirect;
  }

  /// HLS — match `.m3u8` anywhere (path, redirect targets, signed URLs).
  static bool hasObviousM3u8(String url) {
    final u = url.toLowerCase();
    return u.contains('.m3u8');
  }

  /// DASH — match `.mpd` anywhere (manifest URLs, Widevine/ClearKey hosts).
  static bool hasObviousMpd(String url) {
    final u = url.toLowerCase();
    return u.contains('.mpd');
  }

  /// Direct-ish progressive / obvious stream hints when extension is nonstandard.
  static bool hasObviousTs(String url) {
    final u = url.toLowerCase();
    return u.contains('.ts?') || u.endsWith('.ts') || u.contains('.mp4?') || u.endsWith('.mp4');
  }
}

/// Same UA as [playbackHttpHeaders] / native ExoPlayer (real Chrome mobile — avoids bot checks).
const String kBrowserPlaybackUserAgent =
    'Mozilla/5.0 (Linux; Android 13; Mobile) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Mobile Safari/537.36';

File: android/app/src/main/kotlin/com/ayubu/supasoka/player/StreamUrlClassifier.kt
package com.ayubu.supasoka.player

/**
 * Shared gateway / direct-stream URL rules (aligned with Flutter).
 */
object StreamUrlClassifier {

    fun needsWebPlayer(url: String): Boolean {
        val u = url.trim()
        if (u.isEmpty()) return false
        if (hasObviousM3u8(u) || hasObviousMpd(u) || hasObviousTs(u)) return false
        if (isLikelyIptvLiveUrl(u)) return false
        return isPhpLikeUrl(u) || isLikelyGatewayUrl(u)
    }

    fun hasObviousM3u8(url: String) = url.lowercase().contains(".m3u8")

    fun hasObviousMpd(url: String) = url.lowercase().contains(".mpd")

    fun hasObviousTs(url: String): Boolean {
        val l = url.lowercase()
        return l.contains(".ts?") || l.endsWith(".ts") || l.contains(".mp4?") || l.endsWith(".mp4")
    }

    fun hasObviousProgressiveExtension(url: String): Boolean {
        val u = url.lowercase()
        return listOf(".mp4", ".m4v", ".webm", ".mkv", ".mov", ".ts").any { u.contains(it) }
    }

    fun isPhpLikeUrl(url: String?): Boolean {
        if (url.isNullOrBlank()) return false
        return """\.php($|[/?#])""".toRegex(RegexOption.IGNORE_CASE).containsMatchIn(url)
    }

    fun isLikelyGatewayUrl(url: String?): Boolean {
        if (url.isNullOrBlank()) return false
        val u = url.lowercase()
        return """\.(php|asp|aspx|cgi|jsp)(\?|#|$|/)""".toRegex(RegexOption.IGNORE_CASE).containsMatchIn(u) ||
            u.contains("/embed/") ||
            u.contains("/gateway/") ||
            u.contains("/stream/") ||
            u.contains("/play/") ||
            u.contains("/player/")
    }

    fun isRelayStyleUrl(url: String): Boolean {
        val u = url.lowercase()
        return "/relay/stream" in u || "/relay/m3u8" in u || "/api/relay/" in u
    }

    fun isLikelyIptvLiveUrl(url: String): Boolean {
        val base = url.split("#").first().lowercase()
        if (hasObviousM3u8(url) || hasObviousMpd(url) || hasObviousTs(url)) return false
        if ("""^https?://[^/]+:\d{2,5}/(live|stream|play|hls|iptv|channel|ch)/""".toRegex().containsMatchIn(base)) {
            return true
        }
        if ("""^https?://[^/]+:\d{2,5}/[^/]+/[^/]+/[^/?#]+$""".toRegex().containsMatchIn(base)) return true
        if ("""^https?://[^/]+/(live|stream|play|hls|iptv|channel|ch)/[^/?#]+""".toRegex().containsMatchIn(base)) {
            return true
        }
        return false
    }
}
